addons/google_account/models/google_service.py

```diff
@@ -134,32 +135,46 @@ def _do_request(self, uri, params=None, headers=None, method='POST', preuri="htt
-            elif method.upper() in ('POST', 'PATCH', 'PUT'):
-                res = requests.request(method.lower(), preuri + uri, data=params, headers=headers, timeout=timeout)
-            else:
-                raise Exception(_('Method not supported [%s] not in [GET, POST, PUT, PATCH or DELETE]!') % (method))
-            res.raise_for_status()
-            status = res.status_code
-
-            if int(status) in (204, 404):  # Page not found, no response
-                response = False
-            else:
-                response = res.json()
-
+        # Implement exponential backoff as suggested by google documentation
+        # see https://developers.google.com/calendar/v3/errors#403_rate_limit_exceeded
+        retry = 1
+        while True:
+            ask_time = fields.Datetime.now()
             try:
-                ask_time = datetime.strptime(res.headers.get('date'), "%a, %d %b %Y %H:%M:%S %Z")
-            except:
-                pass
-        except requests.HTTPError as error:
-            if error.response.status_code in (204, 404):
-                status = error.response.status_code
-                response = ""
-            else:
-                _logger.exception("Bad google request : %s !", error.response.content)
-                raise error
+                if method.upper() in ('GET', 'DELETE'):
+                    res = requests.request(method.lower(), preuri + uri, params=params, timeout=timeout)
+                elif method.upper() in ('POST', 'PATCH', 'PUT'):
+                    res = requests.request(method.lower(), preuri + uri, data=params, headers=headers, timeout=timeout)
+                else:
+                    raise Exception(_('Method not supported [%s] not in [GET, POST, PUT, PATCH or DELETE]!') % (method))
+                res.raise_for_status()
+                status = res.status_code
+
+                if int(status) in (204, 404):  # Page not found, no response
+                    response = False
+                else:
+                    response = res.json()
+
+                try:
+                    ask_time = datetime.strptime(res.headers.get('date'), "%a, %d %b %Y %H:%M:%S %Z")
+                except:
+                    pass
+                break
+            except requests.HTTPError as error:
+                if error.response.status_code in (204, 404):
+                    status = error.response.status_code
+                    response = ""
+                    break
+                elif error.response.status_code == 403 and any(
+                        e['reason'] == 'rateLimitExceeded'
+                        for e in error.response.json()['error']['errors']):
+                    _logger.info("Rate limit exceeded for google request waiting %s second(s)", retry)
+                    # Max 3 retries, in most experienced cases waiting 1 second and retry is enough
+                    if retry > 4:
+                        raise error
+                    sleep(retry)
```

Comment:
`sleep()` in the middle of a request? ouch! :grimacing: 

exponential backoff is fine if you're doing background work, but it's awful when handling a live request, as it will block a worker, and that can lead to accidental denial of services.

can't we just ignore the error and let it re-sync later?

if we can't, we should distinguish background/cron sync from live requests somehow (without changing the signature, ideally ;-))

Reply:
@odony thx for the feedback, the possibility to dos odoo didn't come into my mind, we can cleary resynchronize later so I will remove the exponential backoff